### ISSUE TITLE
Issues with attrs.setters.NO_OP documentation

### ISSUE THREAD

**Issue by IRDonch** [python-attrs/attrs#917]
There are a few problems with the documentation of this object.

First, [its own docstring](https://www.attrs.org/en/stable/api.html#attrs.setters.NO_OP) is missing, and `object`'s docstring is shown instead.

Second, it's unclear from the documentation of the `on_setattr` parameter of `attr.s`  whether `NO_OP` is allowed there. It just claims that a callable is required, and `NO_OP` isn't one (incidentally, it also doesn't explain what the default of `None` does, which isn't a callable either). However, in the documentation for `attrs.validate` there's an example which calls `define` with `on_setattr=attrs.setters.NO_OP`, which suggests that it _is_ supported.

**Comment by zhiltsov-max** [comment]
Also, is it possible to use "plain" `setattr` (without side-effects) just in the user-provided constructor? The use case is that we'd like to have a more performant implementation of `__init__` (with less function calls, because they start to hit), but still support checks on assignments after initialization.

<details>

```python
def myconv(v):
    print('conv')
    return v

def myvalid(a, b, v):
    print('valid')

@define
class A:
    x = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self):
        self.x = 10

    def f(self):
        print(self.x)

@define(on_setattr=NO_OP)
class B:
    x = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self):
        self.x = 10

    def f(self):
        print(self.x)

@attrs(init=False)
class C:
    x = field(default=5, converter=myconv, validator=myvalid)

    def __init__(self):
        self.x = 10

    def f(self):
        print(self.x)

print('A1')
A()

print('A2')
A().x = 7

print('B1')
B()

print('B2')
B().x = 7

print('C1')
C()

print('C2')
C().x = 7

>>>
A1
conv
valid

A2
conv
valid
conv
valid

B1
B2

C1
C2
```

</details>

**Comment by hynek** [comment]
> Also, is it possible to use "plain" `setattr` (without side-effects) just in the user-provided constructor? The use case is that we'd like to have a more performant implementation of `__init__` (with less function calls, because they start to hit), but still support checks on assignments after initialization.

The generated `__init__` already _does_ use `object.__setattr__`, however it also runs the validators by hand. This is the original behavior. :)

Someone asked about disabling validators before but it seems a weird feature; I think a thread-safe version of `with attrs.validators.disabled():` would be nicer, such that you can instantiate with validated data anytime (#187).

That said, I think using immutable classes might be a better approach.

Other than that, pls don't hijack other issues, this might become hard to follow. Feel free to open an own issue any time!

**Comment by hynek** [comment]
> First, [its own docstring](https://www.attrs.org/en/stable/api.html#attrs.setters.NO_OP) is missing, and `object`'s docstring is shown instead.

I wanted to say that this is embarrassing, but then I checked and it seems like Sphinx changed the behavior of `autodata`  under our butts. :( Check out the 21.2.0 version https://www.attrs.org/en/21.2.0/api.html#attr.setters.NO_OP

I'll have to do some research, but I think I'll just copy paste it and push a post1 tag.

> Second, it's unclear from the documentation of the `on_setattr` parameter of `attr.s` whether `NO_OP` is allowed there.

Ugh, the type is in in the [attr.ib](https://www.attrs.org/en/stable/api.html#attr.ib) definition, but I forgot to add it to attr.s too. Fixed in https://github.com/python-attrs/attrs/commit/fe9432ddf733c736f200b39a8b009bfef4795334.
